include/deal.II/sundials/sunlinsol_newempty.h

```diff
@@ -0,0 +1,146 @@
+ * SUNDIALS package, see the license below.
+ */
+
+/* -----------------------------------------------------------------
+ * Programmer(s): Daniel Reynolds @ SMU
+ *                David J. Gardner, Carol S. Woodward, and
+ *                Slaven Peles @ LLNL
+ * -----------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2002-2020, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * All rights reserved.
+
+  BSD 3-Clause License
+
+  Copyright (c) 2002-2019, Lawrence Livermore National Security and Southern
+  Methodist University. All rights reserved.
+
+  Redistribution and use in source and binary forms, with or without
+  modification, are permitted provided that the following conditions are met:
+
+  * Redistributions of source code must retain the above copyright notice, this
+    list of conditions and the following disclaimer.
+
+  * Redistributions in binary form must reproduce the above copyright notice,
+    this list of conditions and the following disclaimer in the documentation
+    and/or other materials provided with the distribution.
+
+  * Neither the name of the copyright holder nor the names of its
+    contributors may be used to endorse or promote products derived from
+    this software without specific prior written permission.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * -----------------------------------------------------------------*/
+
+#ifndef dealii_sundials_sunlinsol_newempty_h
+#define dealii_sundials_sunlinsol_newempty_h
+
+#include <deal.II/base/config.h>
+#ifdef DEAL_II_WITH_SUNDIALS
+
+DEAL_II_NAMESPACE_OPEN
+namespace SUNDIALS
+{
+  namespace internal
+  {
+    /* -----------------------------------------------------------------
+     * Create a new empty SUNLinearSolver object
+     * ----------------------------------------------------------------- */
+    SUNLinearSolver
+    SUNLinSolNewEmpty()
+    {
+      SUNLinearSolver     LS;
+      SUNLinearSolver_Ops ops;
+
+      /* create linear solver object */
+      LS = NULL;
+      LS = (SUNLinearSolver)malloc(sizeof *LS);
```

Comment:
I assume `SUNLinearSolver` is a pointer type. I think we could use C++ functionality here and instead use `LS = new SUNLinearSolver;` - while we still need to rely on SUNDIALS to properly clean up things and call `SUNLinSolFreeEmpty`, which I do not immediately see how to avoid, we at least avoid some pointer casts discouraged in C++.